### ISSUE TITLE
ICE on chained ref-returning opIndex op-assign

### ISSUE THREAD

**Issue by Temtaime** [ldc-developers/ldc#2415]
```

struct Matrix(uint _M)
{
	enum
	{
		M = _M,
		isVector = M == 1
	}

	union
	{
		float[] flat ;
		float[]A;
	}

	static if(isVector)
		ref opIndex(size_t i) { return flat[i]; }
	else
		ref opIndex(size_t ) 		{
			return *cast(Matrix!(1) *)A;
		}

}

alias Matrix4 = Matrix!(4);

void draw()
{
	Matrix4 mat;

	auto k = mat ;
	k[1][2] += 0.;
}
```

ldc2 crashes
dmd works
OS: windows

**Comment by kinke** [comment]
Thanks, that's an interesting bug. Apparently the lhs matrix element being incremented loses its lvalue-ness after casting from float to double, which makes absolute sense to me. The front-end rewrites `k[1][2] += 0.`, with `k` being a matrix of floats, to `((ref Matrix!(float, 1u) __dop2 = k.opIndex(1LU);) , cast(double)__dop2.opIndex(2LU)) += 0.000000 @ float`, which seems clearly wrong.

As workaround, you can use `k[1][2] += 0.0f;` instead.

**Comment by Temtaime** [comment]
A little more usable version

```
struct Matrix(uint M, uint N = M)
{
	union
	{
		float[M * N] flat;
		float[N][M] A;
	}

	static if(N == 1)
	{
		ref opIndex(size_t i) { return flat[i]; }
	}
	else
	{
		ref opIndex(size_t i) { return *cast(Matrix!(M, 1) *)A[i].ptr; }
	}
}

void main()
{
	Matrix!4 mat;
	mat[1][2] += 0.;
}
```

**Comment by kinke** [comment]
Wrt. the original issue, LDC handles that binassign quirk in the AST wrt. lhs lvalue casting, but it doesn't handle that inside a comma expression. After a quick glance, neither does DMD (they also [skip over casts](https://github.com/dlang/dmd/blob/master/src/ddmd/e2ir.d#L2131-L2159), but don't handle comma expressions in a special way afaict).

Wrt. the arrayop warnings, those shouldn't be reported, so yes, please post a new issue with testcase.

**Comment by kinke** [comment]
DMD seems to handle the comma expressions in `toAddressElem()`; we don't have that option as our `DValue` results of `toElem()` calls don't feature nested DValues as DMD's `elem` structs.

I think the most feasible option for us is to modify the front-end like this:
```
((ref Matrix!(float, 1u) __dop2 = k.opIndex(1LU);) , cast(double)__dop2.opIndex(2LU)) += 0.000000 @ float
=>
cast(double)((ref Matrix!(float, 1u) __dop2 = k.opIndex(1LU);) , __dop2.opIndex(2LU)) += 0.000000 @ float
```

**Comment by kinke** [comment]
Reduced testcase:

```D
struct Row
{
    float val;
    ref float opIndex(int i) { return val; }
}

struct Matrix
{
    Row row;
    ref Row opIndex(int i) { return row; }
}

void main()
{
    Matrix matrix;
    matrix[1][2] += 0.0;
}
```

**Comment by kinke** [comment]
Fixed in #2418.

**Comment by JohanEngelen** [comment]
This introduced regression https://github.com/ldc-developers/ldc/issues/2471.
New fix being worked on: https://github.com/ldc-developers/ldc/pull/2479
